state/watcher.go

```diff
@@ -496,6 +497,186 @@ func watchApplicationRelations(backend modelBackend, applicationName string) Str
 	return newRelationLifeSuspendedWatcher(backend, members, filter, nil)
 }
 
+// WatchModelMachineStartTimes watches the non-container machines in the model
+// for changes to the Life or AgentStartTime fields and reports them as a batch
+// after the specified quiesceInterval time has passed without seeing any new
+// change events.
+func (st *State) WatchModelMachineStartTimes(quiesceInterval time.Duration) StringsWatcher {
+	return newModelMachineStartTimeWatcher(st, st.clock(), quiesceInterval)
+}
+
+type modelMachineStartTimeFieldDoc struct {
+	Id             string    `bson:"_id"`
+	Life           Life      `bson:"life"`
+	AgentStartedAt time.Time `bson:"agent-started-at"`
+}
+
+type modelMachineStartTimeWatcher struct {
+	commonWatcher
+	outCh chan []string
+
+	clk             clock.Clock
+	quiesceInterval time.Duration
+	seenDocs        map[string]modelMachineStartTimeFieldDoc
+}
+
+func newModelMachineStartTimeWatcher(backend modelBackend, clk clock.Clock, quiesceInterval time.Duration) StringsWatcher {
+	w := &modelMachineStartTimeWatcher{
+		commonWatcher:   newCommonWatcher(backend),
+		outCh:           make(chan []string),
+		clk:             clk,
+		quiesceInterval: quiesceInterval,
+		seenDocs:        make(map[string]modelMachineStartTimeFieldDoc),
+	}
+
+	w.tomb.Go(func() error {
+		defer close(w.outCh)
+		return w.loop()
+	})
+	return w
+}
+
+// Changes returns the event channel for the watcher.
+func (w *modelMachineStartTimeWatcher) Changes() <-chan []string {
+	return w.outCh
+}
+
+func (w *modelMachineStartTimeWatcher) loop() error {
+	docWatcher := newCollectionWatcher(w.backend, colWCfg{col: machinesC})
+	defer func() { _ = docWatcher.Stop() }()
+
+	var (
+		timer           = w.clk.NewTimer(w.quiesceInterval)
+		timerArmed      = true
+		unprocessedDocs = make(set.Strings)
+		outCh           chan []string
+		changeSet       []string
+	)
+	defer func() { _ = timer.Stop() }()
+
+	// Collect and initial set of machine IDs; this makes the worker
+	// compatible with other workers that expect the full state to be
+	// immediately emitted once the worker starts.
+	initialSet, err := w.initialMachineSet()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	changeSet = initialSet.Values()
+	outCh = w.outCh
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-w.watcher.Dead():
+			return stateWatcherDeadError(w.watcher.Err())
+		case changes := <-docWatcher.Changes():
```

Comment:
Given that we are watching all machines in the collection, is it worth it to filter out containers by their ID (which always has a '/' in it) rather than aggregating the container ids, and then filtering it out in the processChanges?

I'm just thinking that a lot of our deploys are actually dominated by containers (default openstack bundle uses containers for all openstack control plane)

Looking at 'machineDoc' it doesn't look like many of those fields are going to be changing frequently. They'll all change a couple times during initialization (addresses, primary units, etc), but they shouldn't be continually changing in a stable system.
It is a bit of a shame that changing a machine's address is going to wake us up here, but I guess we at least don't cause the watcher of this to wake up again.

Reply:
Good idea. I will filter the containers out at this stage.